agent-control/src/opamp/attributes.rs

```diff
@@ -0,0 +1,222 @@
+use std::error::Error;
+use std::fmt::Debug;
+
+use opamp_client::StartedClient;
+use opamp_client::opamp::proto::any_value::Value;
+use opamp_client::opamp::proto::{AgentDescription, AnyValue, KeyValue};
+use tracing::error;
+
+use crate::event::channel::EventPublisher;
+
+/// Event message type for updating OpAMP agent attributes
+pub type UpdateAttributesMessage = Vec<Attribute>;
+
+/// Represents the type of an [AgentDescription message in OpAMP](https://opentelemetry.io/docs/specs/opamp/#agentdescription-message)
+#[derive(Debug, Clone, PartialEq)]
+pub enum AttributeType {
+    Identifying,
+    NonIdentifying,
+}
+
+/// Represents an agent attribute
+///
+/// Simple wrapper for [`KeyValue`] and it's type, that simplifies the creation
+/// of agent attributes.
+///
+/// ## Example:
+/// ```
+/// # use newrelic_agent_control::opamp::attributes::{Attribute, AttributeType};
+/// let attr = Attribute::from((AttributeType::Identifying, "key", "value"));
+/// ```
+/// instead of (assume we don't care about the type for this example):
+/// ```
+/// # use opamp_client::opamp::proto::any_value::Value;
+/// # use opamp_client::opamp::proto::{AnyValue, KeyValue};
+/// let attr = KeyValue {
+///     key: "key".into(),
+///     value: Some(AnyValue {
+///         value: Some(Value::StringValue("value".into())),
+///     }),
+/// };
+/// ```
+#[derive(Debug, Clone, PartialEq)]
+pub struct Attribute {
+    attribute_type: AttributeType,
+    key_value: KeyValue,
+}
+
+impl Attribute {
+    /// Returns the inner [`KeyValue`]
+    pub fn key_value(self) -> KeyValue {
+        self.key_value
+    }
+}
+
+impl<K, V> From<(AttributeType, K, V)> for Attribute
+where
+    K: AsRef<str>,
+    V: AsRef<str>,
+{
+    fn from((attribute_type, key, value): (AttributeType, K, V)) -> Self {
+        Attribute {
+            attribute_type,
+            key_value: KeyValue {
+                key: String::from(key.as_ref()),
+                value: Some(AnyValue {
+                    value: Some(Value::StringValue(String::from(value.as_ref()))),
+                }),
+            },
+        }
+    }
+}
+
+/// Updates the attributes of the OpAMP agent
+///
+/// If an attribute already exists, it will be updated. If it doesn't, it will be added.
+/// If the `new_attributes` contain duplicated keys, the last occurrence will be kept.
+pub fn update_opamp_attributes<C>(
+    opamp_client: &C,
+    new_attributes: Vec<Attribute>,
+) -> Result<(), Box<dyn Error>>
```

Comment:
Nit: should we return `opamp_client::client::ClientError` here?

Reply:
I don't know. It doesn't matter. We just print it in the callers.

https://github.com/newrelic/newrelic-agent-control/pull/1956/changes/42b95450a501c3102c09c01b7874b305229632d4